### PR TITLE
[BEAM-831] ParDo Chaining of Apex Runner

### PR DESCRIPTION
**Pull request by chinmaykolhatkar**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

R: @tweise, @kennknowles, @jkff 
---

This is currently Work in Progress PR. The tests are being developed. This PR was created to have a convenient environment for discussing this code. Here are the TODOS:
1. Create the test cases for ParDo fusion for Apex runner

### CODE REVIEW

File: runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java [hunk 1]
```diff
@@ -62,4 +62,9 @@
   @Default.String("classpath:/beam-runners-apex.properties")
   String getConfigFile();
 
+  @Description("configure whether to perform ParDo fusion")
```
Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java]: I don't think this is needed. It should always be done. The user can tweak stream locality through the apex properties if needed.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java [hunk 2]
```diff
@@ -142,16 +147,19 @@ public void addOperator(Operator operator, OutputPort port, PCollection output)
...
   public void addStream(PInput input, InputPort inputPort) {
     while (aliasCollections.containsKey(input)) {
       input = aliasCollections.get(input);
     }
-    Pair<OutputPort<?>, List<InputPort<?>>> stream = this.streams.get(input);
-    checkArgument(stream != null, "no upstream operator defined for %s", input);
-    stream.getRight().add(inputPort);
+
+    Pair<OutputPortInfo, List<InputPortInfo>> stream = this.streams.get(input);
+    checkArgument(stream != null, "no upstream operator defined for " + input);
```
Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java]: why this change?

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java [hunk 3]
```diff
@@ -191,11 +209,74 @@ public void populateDAG(DAG dag) {
     }
   }
 
+  private void optimizeStreams(DAG.StreamMeta streamMeta,
```
Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java]: A description what this does is indicated.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java [hunk 4]
```diff
@@ -191,11 +209,74 @@ public void populateDAG(DAG dag) {
...
+          ParDo.BoundMulti t = (ParDo.BoundMulti) transform;
+          if (t.getSideInputs().size() > 0) {
+            loc = DAG.Locality.CONTAINER_LOCAL;
+            break;
+          } else {
+            loc = DAG.Locality.THREAD_LOCAL;
+          }
+        } else {
+          // Sink is not ParDo.. set null locality.
+          loc = null;
```
Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java]: Isn't that the default already?

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java [hunk 5]
```diff
@@ -191,11 +209,74 @@ public void populateDAG(DAG dag) {
...
+    PTransform sourceTransform = source.transform.getTransform();
+    if (sourceTransform instanceof ParDo.Bound
+        || sourceTransform instanceof ParDo.BoundMulti) {
+      // Source is ParDo.. Check sink(s)
+      for (InputPortInfo sink : sinks) {
+        PTransform transform = sink.transform.getTransform();
+        if (transform instanceof ParDo.Bound) {
+          ParDo.Bound t = (ParDo.Bound) transform;
+          if (t.getSideInputs().size() > 0) {
+            loc = DAG.Locality.CONTAINER_LOCAL;
```
Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java]: What if side inputs are used by multiple transforms that are not part of a single fused chain? Then they should not be forced in the same container.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java (hunk 1 above)
Comment (chinmaykolhatkar) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java]: Done. Removed the pipeline option.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java (hunk 2 above)
Comment (chinmaykolhatkar) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java]: YARN launcher was failing at this change... I think in my copy of beam. Now it is fine. Reverted this change.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java (hunk 4 above)
Comment (chinmaykolhatkar) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java]: Yes.. Just added for completion. I can remove it.

Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/translation/TranslationContext.java]: If it isn't needed, then please remove it.

**Comment by tweise** [comment]
@chinmaykolhatkar build shows a failure. Also can you use the word count example to show how your changes have reduced the number of containers.

**Comment by tweise** [comment]
Also needs a rebase after BEAM-1709 and please address other comments.

**Comment by tweise** [comment]
@chinmaykolhatkar any update on this? What does it take to get this to the finish line (it is essential for the runner)? You would also need to rebase and there is now another operator that needs to be fused: ApexProcessFnOperator

**Comment by chinmaykolhatkar** [comment]
@tweise Apologies for delay. I'm still not able to get the through my task list to reach this one. I'll try to do this sooner.

**Comment by kennknowles** [comment]
It would be really nice to have this for the first stable release, so that users have a better experience with the Apex runner right away.
